pkg/kubelet/cm/topologymanager/topology_manager.go

```diff
@@ -54,3 +67,217 @@ type TopologyHint struct {
+	pth := make(map[string]map[string]TopologyHint)
+	pm := make(map[string]string)
+	manager := &manager{
+		hintProviders:    hp,
+		podTopologyHints: pth,
+		podMap:           pm,
+		policy:           policy,
+	}
+
+	return manager
+}
+
+func (m *manager) GetAffinity(podUID string, containerName string) TopologyHint {
+	return m.podTopologyHints[podUID][containerName]
+}
+
+// Iterate over all permutations of hints in 'allProviderHints [][]TopologyHint'.
+//
+// This procedure is implemented as a recursive function over the set of hints
+// in 'allproviderHints[i]'. It applies the function 'callback' to each
+// permutation as it is found. It is the equivalent of:
+//
+// for i := 0; i < len(providerHints[0]); i++
+//     for j := 0; j < len(providerHints[1]); j++
+//         for k := 0; k < len(providerHints[2]); k++
+//             ...
+//             for z := 0; z < len(providerHints[-1]); z++
+//                 permutation := []TopologyHint{
+//                     providerHints[0][i],
+//                     providerHints[1][j],
+//                     providerHints[2][k],
+//                     ...
+//                     provideryHints[-1][z]
+//                 }
+//                 callback(permutation)
+func (m *manager) iterateAllProviderTopologyHints(allProviderHints [][]TopologyHint, callback func([]TopologyHint)) {
+	// Internal helper function to accumulate the permutation before calling the callback.
+	var iterate func(i int, accum []TopologyHint)
+	iterate = func(i int, accum []TopologyHint) {
+		// Base case: we have looped through all providers and have a full permutation.
+		if i == len(allProviderHints) {
+			callback(accum)
+			return
+		}
+
+		// Loop through all hints for provider 'i', and recurse to build the
+		// the permutation of this hint with all hints from providers 'i++'.
+		for j := range allProviderHints[i] {
+			iterate(i+1, append(accum, allProviderHints[i][j]))
+		}
+	}
+	iterate(0, []TopologyHint{})
+}
+
+// Merge the hints from all hint providers to find the best one.
+func (m *manager) calculateAffinity(pod v1.Pod, container v1.Container) TopologyHint {
+	// Set the default hint to return from this function as an any-socket
+	// affinity with an unpreferred allocation. This will only be returned if
+	// no better hint can be found when merging hints from each hint provider.
+	defaultAffinity, _ := socketmask.NewSocketMask()
+	defaultAffinity.Fill()
+	defaultHint := TopologyHint{defaultAffinity, false}
+
+	// Loop through all hint providers and save an accumulated list of the
+	// hints returned by each hint provider. If no hints are provided, assume
+	// that provider has no preference for topology-aware allocation.
+	var allProviderHints [][]TopologyHint
+	for _, provider := range m.hintProviders {
+		// Get the TopologyHints from a provider.
+		hints := provider.GetTopologyHints(pod, container)
+
+		// If hints is nil, overwrite 'hints' with a preferred any-socket affinity.
+		if hints == nil {
```

Comment:
This case should handle empty slice as well (empty initialized slice does not compare equal to nil). Without that check, any provider who returns an empty slice effectively short-circuits all of the following logic.

Reply:
Hmm. I wrote this logic assuming that `nil` was the only way to indicate "don't care" as the return from `GetTopologyHints()`.

If, instead, `[]TopologyHints{}` is returned as an empty list, I think that should be interpreted as "caring", but having no current way of doing exclusive alignment (equivalent to a single hint of {1,1 false} on a 2 socket system).

Looking deeper, this is not what the code is currently doing, but I think we should change the logic to enforce this.
i.e. adding logic directly below this as:
```
// If hints is an empty list, overwrite 'hints' with a non-preferred any-socket affinity.
if len(hints) == 0 {
    affinity, _ := socketmask.NewSocketMask()
    affinity.Fill()
    hints = []TopologyHint{{affinity, false}}
}
```
Thoughts?

Reply:
Or is this maybe too subtle. It means that the following loops will result in different behaviour:
```
var hints []TopologyHint
for {
    // populate hints if possible
}
return hints

hints := []TopologyHint{}
for {
    // populate hints if possible
}
return hints
```